exchange/targeting_test.go

```diff
@@ -247,3 +247,166 @@ func (f *mockFetcher) GetId(bidder openrtb_ext.BidderName) (string, bool) {
 func mockServer(w http.ResponseWriter, req *http.Request) {
 	w.Write([]byte("{}"))
 }
+
+type TargetingTestData struct {
+	Description                string
+	TargetData                 targetData
+	Auction                    auction
+	IsApp                      bool
+	CategoryMapping            map[string]string
+	ExpectedBidTargetsByBidder map[string]map[openrtb_ext.BidderName]map[string]string
+}
+
+func TestSetTargeting(t *testing.T) {
+	for _, test := range TargetingTests {
+		auc := &test.Auction
+		// Set rounded prices from the auction data
+		auc.setRoundedPrices(test.TargetData.priceGranularity)
+		winningBids := make(map[string]*pbsOrtbBid)
+		// Set winning bids from the auction data
+		for imp, bidsByBidder := range auc.winningBidsByBidder {
+			for _, bid := range bidsByBidder {
+				if winningBid, ok := winningBids[imp]; ok {
+					if winningBid.bid.Price < bid.bid.Price {
+						winningBids[imp] = bid
+					}
+				} else {
+					winningBids[imp] = bid
+				}
+			}
+		}
+		auc.winningBids = winningBids
+		targData := test.TargetData
+		targData.setTargeting(auc, test.IsApp, test.CategoryMapping)
+		for imp, targetsByBidder := range test.ExpectedBidTargetsByBidder {
+			for bidder, expected := range targetsByBidder {
+				assert.Equal(t,
+					expected,
+					auc.winningBidsByBidder[imp][bidder].bidTargets,
+					"Test: %s\nTargeting failed for bidder %s on imp %s.",
+					test.Description,
+					string(bidder),
+					imp)
+			}
+		}
+	}
+
+}
+
+var TargetingTests []TargetingTestData = []TargetingTestData{
```

Comment:
Some of these tests aren't related to hb_format. Did we have testing gaps before that you're addressing?

Reply:
I believe so. We had no testing on the `setTargeting()` level. I believe the higher level tests would end up covering some cases, but they are rather far removed, so would be a lot of work to confirm which cases are covered at that level. As you can see, `setTargeting()` requires a lot of data to be set up, which is probably why it was not done before. But now with this framework in place we can easily expand it next time we touch targeting or find a specific hole in our testing.

Reply:
Thanks a lot @hhhjort for adding these tests. They are definitely very helpful. Could we maybe also add some test cases around the other targeting keys like the cache ones and the deal ones? I understand that this PR doesn't make any code changes for those but since we have this framework in place now, I am hoping that it wouldn't be too much of an effort to add that. I will leave this up to you though.